### PR TITLE
refund endpoint

### LINKED ISSUE #14: Refunds
In the other team's documentation, they had an option to refund the payments. We could add a simple refund state, where it's assumed that the refund is done manually and after that in the PoS system the payment is marked as refunded.

Technical:
- Add a new payment status `refunded`.
- Create a `PUT /payment/:id/refund` endpoints that would transition the payment into the `refunded` state (only if that payment was in a `completed` state)

### PR DESCRIPTION
**Pull request by RugileVa**
closes #14

### CODE REVIEW

**Comment by 6ar8nas** [comment]
Closes #14

File: src/modules/payment/payment.controller.ts [hunk 1]
```diff
@@ -152,3 +153,31 @@ export const completePaymentHandler =
...
+        .send({ message: `Payment '${params.id}' was not found` });
+
+    if (paymentBefore.status !== PaymentStatus.completed) {
+      const error: ValidationErrorResponse = {
+        message: `Payment '${params.id}' cannot be refunded. It is not in a completed state.`,
+        issues: [],
+      };
+
+      return reply.code(400).send(error);
+    }
```
Comment (6ar8nas) [inline src/modules/payment/payment.controller.ts:179]: The issues array being empty, but the problem is added to the message is a bit counter-intuitive.

Comment (6ar8nas) [reply src/modules/payment/payment.controller.ts:179]: Also, this is not necessarily a validation error. 
Maybe instead we could fetch for completed payments and throw a 404 with a message `Completed payment $id was not found`?

Comment (6ar8nas) [reply src/modules/payment/payment.controller.ts:179]: Another way of solving it _could_ be returning `409 Conflict` with errorResponse schema, as by definition it matches the description.
image

However, I think some of the above options are better.

**Review by 6ar8nas — CHANGES_REQUESTED** [review]

File: src/modules/payment/payment.route.ts [hunk 2]
```diff
@@ -120,5 +121,23 @@ export const paymentRoutes = (server: FastifyInstance) => {
...
+    "/:id/refund",
+    {
+      schema: {
+        tags: ["Payments"],
+        summary: "Refund a payment",
+        params: $genericRef("requestByIdParams"),
+        response: {
+          200: $paymentRef("paymentResponseSchema"),
+          404: $genericRef("errorResponse"),
+          400: $genericRef("validationErrorResponse"),
```
[the remark this answers is not in the document]
Comment (RugileVa) [reply src/modules/payment/payment.route.ts:134]: Swagger show errors in the correct order.

[the remark this answers is not in the document]
Comment (6ar8nas) [reply src/modules/payment/payment.route.ts:134]: Appears to work fine. Resolving.

File: src/modules/payment/payment.route.ts [hunk 3]
```diff
@@ -120,5 +121,23 @@ export const paymentRoutes = (server: FastifyInstance) => {
...
+        },
+      },
+      preValidation: preValidationHandler(requestByIdParams, undefined),
```
[the remark this answers is not in the document]
Comment (6ar8nas) [reply src/modules/payment/payment.route.ts:137]: I can see that this is the way it's been done in payments API. Resolving, no need to change anything for in-file consistency.

File: src/modules/payment/payment.controller.ts (hunk 1 above)
Comment (6ar8nas) [reply src/modules/payment/payment.controller.ts:179]: @BenasB what do you think - should we address this?

**Comment by BenasB** [comment]
closes #14

File: src/modules/payment/payment.controller.ts (hunk 1 above)
Comment (6ar8nas) [reply src/modules/payment/payment.controller.ts:179]: Let's use ErrorResponse and make the ValidationErrorResponse extend the Error response with optional issues.

File: src/modules/payment/payment.controller.ts [hunk 4]
```diff
@@ -152,3 +150,30 @@ export const completePaymentHandler =
...
+  ) => {
+    const { params } = request;
+    const paymentBefore = await findPaymentById(server, params.id);
+    if (!paymentBefore)
+      return reply
+        .code(404)
+        .send({ message: `Payment '${params.id}' was not found` });
+
+    if (paymentBefore.status !== PaymentStatus.completed) {
+      const error: ValidationErrorResponse = {
```
Comment (6ar8nas) [inline src/modules/payment/payment.controller.ts:170]: Simple ErrorResponse instead. Export if needed.

**Review by 6ar8nas — APPROVED** [review]

**Pull request merged**